### DISCUSSION TITLE
How to run ActiveAdmin v4 without node and yarn installed?

### DISCUSSION THREAD

**Discussion by amkisko** [activeadmin/activeadmin#8223] [Help]
importmap-rails and propshaft do not require node and yarn to be installed, although cssbundling-rails requires.

Are there any options to patch ActiveAdmin to have no dependence on node/yarn?

References:
- https://github.com/activeadmin/activeadmin/blob/master/UPGRADING.md

**Comment by javierjulio** [comment]
Hi @amkisko. First and foremost, thank you for creating a discussion. This may be doable from some past experimenting so I will share my thoughts on how I think it could work but it's not something that the library will support. I know some others are interested in this as well.

You would have to use and customize `tailwindcss-rails` [since it is designed for a single entrypoint file](https://github.com/rails/tailwindcss-rails/blob/main/lib/tailwindcss/commands.rb#L76-L85). You could install that gem and start copying it's tasks to work with the `tailwind-active_admin.config.js` file or if you aren't using Tailwind in your app, and only plan to do so with ActiveAdmin, then just update the files to work directly with the `tailwindcss-rails` defaults, assuming your application styles can live in a different named entrypoint file.

From there, you'll have to copy the `plugin.js` file to the root of your project and keep it up to date. Since you don't want to run Node, you will need that file locally because we provide some CSS through a Tailwind plugin and for the standalone CLI to run that plugin, it has to be relative/local. I've tried executing it from the gem path since it's included in the gem but it fails with an error. Then update your config to require the relative file `require("./plugin")` and it may work.

Resources:
- https://github.com/rails/tailwindcss-rails/issues/151
- https://github.com/tailwindlabs/tailwindcss/discussions/7698
- https://github.com/rails/tailwindcss-rails/issues/108

**Reply by amkisko** [reply]
Thanks, sounds promising. First thought was if it is possible to use `importmap-rails` for fetching artefacts out from `tailwindcss-rails`... seems to be not that straight forward thing.

To me it seems that with ActiveAdmin v4 got strong requirement of node due to `cssbundling-rails` only, but previously it was possible to detach from node, using `propshaft` and `importmap-rails` for example. 🤔

Reasons to switch to Tailwind are quite clear.

**Comment by djolereject** [comment]
> ...

Would it be possible to get some kind of example app or at least more detailed solution? I'm trying to setup exactly this, but still struggling and can't quite make it work. What should we put in this config file, is there a need for another tailwind css file besides generated one and where from to call require `plugin`... I'm doing something wrong, but not sure what.
Thanks!

**Comment by henrikbjorn** [comment]
@amkisko @djolereject @javierjulio 

This is what I did as part of my upgrade from ActiveAdmin 3.x to the latest 4.0 (beta.4).

I took inspiration in the `tailwindcss-rails` code and basically used the same pattern, which together with `Procfile.dev` makes my rails setup completely node less.

I think this also could be a good default for ActiveAdmin, since I think many are wanting to not have node installed when developing with Rails (`dartsass-rails`, `tailwindcss-rails`, `mrml-rails` are examples of this)

### How to do it

1. I moved `tailwind-active_admin.config.js` into the `config/` directory. Since the other config from `tailwindcss-rails` is there.
2. Changed the plugin require statement to `require(`${activeAdminPath}/plugin.js`)` since the code is in the gem anyways, and the path is already resolved. This way I don't need to maintain anything separately.
3. Then I ran `bundle binstub tailwindcss` so that I get a `bin/tailwindcss` bin script
4. Dumped the following into `lib/tasks/active_admin.rb` https://gist.github.com/henrikbjorn/cf9ab168bb6ff9195691d7add5cb7d1b It will hook into the precompile tasks (as `tailwindcss-rails`)
5. Into `Procfile.dev` add a line `active_admin: bin/rails active_admin:watch`.
6. 🥳

**Reply by henrikbjorn** [reply]
I made a typo in the rake task, which I have now corrected

**Reply by henrikbjorn** [reply]
Did a quick blog post on it also: https://medium.com/@henrikbjorn/til-how-to-use-activeadmin-v4-without-installing-nodejs-ccdaf3ace005

**Reply by amkisko** [reply]
I started from scratch with app that has v2 and it required these commands:
```
$ bin/rails tailwindcss:install
$ rails generate active_admin:assets
$ mv tailwind-active_admin.config.js config
$ touch lib/tasks/active_admin.rake
$ bundle binstub tailwindcss-rails
```

Fix:
```
- bundle binstub tailwindcss
+ bundle binstub tailwindcss-rails
```

**Reply by javierjulio** [reply]
@thibaudgg you should give this approach a try to see if it works for you.

**Reply by thibaudgg** [reply]
@javierjulio nice, I'll give it a try soon. 👍

**Reply by amkisko** [reply]
This approach worked for me mostly, not all things worked out of box and required tuning here and there, also `Upgrade v3 to v4` notes were very useful, now I am facing a bit unrelated to the topic issues, but still reported here. :)

**Reply by amkisko** [reply]
Collected all steps in single gist here: https://gist.github.com/amkisko/c704c1a6462d573dfa4820ae07d807a6

**Reply by amkisko** [reply]
@henrikbjorn have you figured out how to extend ActiveAdmin importmap with custom scripts from main app?

**Reply by henrikbjorn** [reply]
@amkisko Can't you just do something like:

``` ruby
ActiveAdmin.importmap.draw do
  # do the pin stuff here ?
end
```

Seems to be used in the partial https://github.com/activeadmin/activeadmin/blob/01d0156c40a60fb401c0c31dd07ea8b974116063/app/views/active_admin/_html_head.html.erb#L13

**Reply by amkisko** [reply]
Yes, I saw this one and tried to hook into ActiveAdmin importmap without proper success. Probably it should be done in combination with `javascript_import_module_tag "application"` after `javascript_importmap_tags`. Resolved by fully copying javascript from ActiveAdmin and having custom importmap class.

**Reply by henrikbjorn** [reply]
@amkisko you should be able to just pin stuff into it there. No need for anything custom

**Reply by amkisko** [reply]
Works perfect! Thanks @henrikbjorn!

In `views/active_admin/_html_head.html.erb`:
```
<%= javascript_importmap_tags "active_admin", importmap: ActiveAdmin.importmap %>
<%= javascript_import_module_tag "application" %>
```

In `config/initializers/active_admin.rb`:
```
ActiveAdmin.importmap.draw do
  pin "@rails/actioncable", to: "actioncable.esm.js", preload: true
  pin "@rails/activestorage", to: "activestorage.esm.js", preload: true

  pin "@hotwired/turbo-rails", to: "turbo.js", preload: true
  pin "@hotwired/stimulus", to: "stimulus.js", preload: true
  pin "@hotwired/stimulus-loading", to: "stimulus-loading.js", preload: true

  pin "application", preload: true
  pin_all_from "app/assets/javascripts/controllers", under: "controllers"
end

ActiveAdmin.setup do |config|
  ...
end
```

**Reply by ryanckulp** [reply]
nice work @henrikbjorn ! just a couple things:

1. your docs (Step 4) should say `lib/tasks/active_admin.rake` vs `.rb`
2. the task itself is not super DRY and also does not take advantage of minification; slightly better gist [here](https://gist.github.com/ryanckulp/4872d0379b642421928b9d24d9c5d38c) 

thanks again!

**Reply by RathanKumar** [reply]
If anyone is facing issue while running `bin/dev` after updating Procfile.dev. I noticed that `bundle binstub tailwindcss-rails` doesn't add `tailwindcss` binary into bin. Instead, try `bundle binstub tailwindcss-ruby`

**Comment by djolereject** [comment]
Thank for the detailed explanation!

**Comment by amkisko** [comment]
What is the suggestion on extending header and layout of ActiveAdmin v4?

That's what I been up to before:
```ruby
module AdminPageLayoutOverride
  def build_page(*args)
    within super do
      render "active_admin/layout_additions"
    end
  end

  def build_active_admin_head
    within super do
      render "active_admin/header"
    end
  end
end

ActiveAdmin::Views::Pages::Base.prepend AdminPageLayoutOverride
```

**Reply by henrikbjorn** [reply]
Overwrite the partials that you can get from `bin/rails g active_admin:views`

**Comment by amkisko** [comment]
This thread might become a more general "migration to v4" topic.

I've got another one, that can't figure out how to fix, getting this for all resources except custom pages:
```
No route matches {:action=>"index", :controller=>"active_admin/users", :format=>:csv}
```

```
No route matches {:action=>"index", :controller=>"active_admin/user_metrics"}
```

**Reply by amkisko** [reply]
This one happens during migration from v3 to v4. Does not happen in the project started from scratch.

**Reply by amkisko** [reply]
This one is resolved, details here: https://github.com/activeadmin/activeadmin/issues/8279

**Comment by amkisko** [comment]
Also it was not really clear how `main_app` variable change affects the migration, basically I started using it everywhere for all routes paths, even this one:
```
# PATH: config/initializers/active_admin.rb
ActiveAdmin.setup do |config|
  ...
  config.logout_link_path = ->(*) { main_app.destroy_user_session_path }
  ...
end
```

**Comment by kirillplatonov** [comment]
It would be much easier if you just shipped bundled versions of `active_admin.css`. Having strict dependency on `cssbundling-rails` and `tailwind` to build CSS for third party library in the host app sounds weird. It will make adoption of new version harder. Personally, I very rarely have the need to customize ActiveAdmin CSS and I would rather add regular CSS classes when necessary.

Here's an example of how Avo solves it:
https://avohq.io/blog/how-to-bundle-assets-in-a-rails-engine

But it can be even simpler. Just a pre-commit hook that will run tailwind building and save the output into `app/assets/stylesheets/active_admin.css`.

**Reply by javierjulio** [reply]
@kirillplatonov it would not have been easier. From your perspective but not from the library's. There are more Rails asset management libraries now than before which means more possible combinations of them and others expect each to just work without much effort. That's difficult and time consuming to figure out how to get each implemented (e.g. you may not be familiar with some) and then also having each tested in an engine. It doesn't sound weird at all when you put yourself in the shoes of someone that actually has to do the work.

I've read their approach before I started working on v4 last year. I made a deliberate choice to **not** do that. Compiling the library CSS in advance is a bad choice as styles can clash with the parent app (ordering, colors, etc.). While you may not want to make changes, others do. The approach they shared works well for a library UI that you wouldn't need to make changes. If I recall right, the inspiration for that approach came from Sidekiq. In my research, GoodJob does something similar as well.

**Reply by jaynetics** [reply]
There is a now gem that does this, see [my answer below](https://github.com/activeadmin/activeadmin/discussions/8223#discussioncomment-10042712).

**Comment by amkisko** [comment]
@javierjulio what is the current support level for Turbo in ActiveAdmin? At least I have not seen any traces of usage here. One thing that does not properly work is form errors, there is a reply from server but HTML code from response is not applied to the current page, I've been experiencing such starting from v3. Have you seen such behaviour?

**Reply by javierjulio** [reply]
There is no Turbo support. I'm not familiar with it. I don't know how well it would work with Flowbite.

**Reply by tagliala** [reply]
I'm familiar with Turbo, and there is a PR to improve responses from arbre to support Turbo

However, it is not clear to me what "support turbo" means.

Does it mean:
1. The hosting application uses Turbo, and Active Admin does not work properly; or
2. Active Admin should implement navigation with Turbo; or
3. Something else

?

**Reply by amkisko** [reply]
I meant option 1, there are issues with rendering validation errors within form due to Turbo. I have not yet tested it with fully disabled Turbo, but I bet it will work if disabled.

One more visible issue - navigation sidebar on smaller screens has button to show, it does not work after first navigation click, so basically it loses onclick handler after Turbo page update.

**Reply by amkisko** [reply]
@tagliala where is the mentioned PR? I found at least three closed and not merged PRs here in repository.

**Reply by tagliala** [reply]
> I meant option 1, there are issues with rendering validation errors within form due to Turbo. I have not yet tested it with fully disabled Turbo, but I bet it will work if disabled.

You mean if you navigate to `/admin` from the app?

**Reply by tagliala** [reply]
> @tagliala where is the mentioned PR? I found at least three closed and not merged PRs here in repository.

https://github.com/activeadmin/inherited_resources/pull/863

**Reply by amkisko** [reply]
> You mean if you navigate to /admin from the app?
When already working with ActiveAdmin and actively navigating, Turbo breaks listeners on ActiveAdmin elements.

<img width="730" alt="Screenshot 2024-05-10 at 10 45 00" src="https://github.com/activeadmin/activeadmin/assets/54796266/1df399d6-2337-4f8a-b68f-ca4d49588f46">

Reproducible here: https://github.com/kiskolabs/activeadmin-demo

**Reply by tagliala** [reply]
Page not found, is that a private repo?

**Reply by amkisko** [reply]
On form save it produces `Error: Form responses must redirect to another location`
<img width="1217" alt="Screenshot 2024-05-10 at 10 56 01" src="https://github.com/activeadmin/activeadmin/assets/54796266/e5c1ec67-22d1-49c8-96ad-019d54380378">

**Reply by amkisko** [reply]
> Page not found, is that a private repo?

Ah, sorry, will make it public soon!

**Reply by amkisko** [reply]
Fixed, repo is now available for everyone.

**Reply by tagliala** [reply]
I can imagine that there are use cases for this, but I did not expect the following:

https://github.com/kiskolabs/activeadmin-demo/blob/b0641022af71b679f6435479f6c4bc0109ac092d/app/views/active_admin/_html_head.html.erb#L14

@javierjulio is it a good practice or common practice to include application javascript in active admin?

**Reply by amkisko** [reply]
I can rename it to `custom_active_admin` if that's critical. :)

**Reply by amkisko** [reply]
So I decided to go with `<body data-turbo="false"` for ActiveAdmin layout.

**Reply by hoquandai** [reply]
Hi @amkisko,
ActiveAdmin is using Flowbite.js version which is not compatible with Turbo, [code line](https://github.com/activeadmin/activeadmin/blob/v4.0.0.beta7/config/importmap.rb#L2).
Therefore, you should consider to override the pinned package with the new one played well with Turbo, [here](https://flowbite.com/docs/getting-started/rails/#importmap).

```ruby
# active_admin.rb

### Example:

########### ACTIVEADMIN IMPORTMAP ###########

ActiveAdmin.after_load do
  ActiveAdmin.importmap.draw do
    pin "application"
    pin "flowbite", to: "flowbite.turbo.min.js" # <- Override flowpite
    pin "@hotwired/turbo-rails", to: "turbo.min.js"
    pin "@hotwired/stimulus", to: "stimulus.min.js"
    pin "@hotwired/stimulus-loading", to: "stimulus-loading.js"
    pin_all_from "app/javascript/controllers", under: "controllers"
  end
end

############################################
```

**Reply by amkisko** [reply]
Thanks @hoquandai, probably this fix even eligible for Pull Request here? I have not thought in details and I am already aware that ActiveAdmin v4 works well with Rails 6 (no Turbo), so there might be some things that should be considered for such changes.

**Reply by jsntv200** [reply]
The following has been working for me with the benefit of retaining rails default behaviour for pinning and avoids duplication:

```ruby
ActiveAdmin.after_load do
  ActiveAdmin.importmap.draw(Rails.root.join("config/importmap.rb"))
  ActiveAdmin.importmap.cache_sweeper(watches: [ Rails.root.join("app/javascript") ])
end
```

As a side note there's currenlty 2 open PR's for import-map rails to handle multiple imports:
https://github.com/rails/importmap-rails/pull/241
https://github.com/rails/importmap-rails/pull/253

**Reply by amkisko** [reply]
@jsntv200 thanks for sharing, one obvious problem is that all the pins from default importmap.rb will leak to activeadmin side too, sometimes it makes sense to split them, especially when there are possible naming conflicts.

**Comment by jaynetics** [comment]
I've made a gem to allow running ActiveAdmin v4 without any asset setup: [activeadmin_assets](https://github.com/jaynetics/activeadmin_assets/). Please note that it limits customizability as described in the README.

**Reply by henrikbjorn** [reply]
Smart!

**Reply by atsjj** [reply]
This is excellent! Thanks for the gem!

**Comment by asgeo1** [comment]
I wanted to get ActiveAdmin v4 (beta13) working with Rails v8.0.0, using `tailwindcss-rails` and `importmaps-rails` - so I didn't have to run a separate node / npm process.

In my case I wasn't using tailwind for anything else, and rails is mostly being used just as an API.

This is what I had to do:

- install `activeadmin`, `tailwindcss-rails` and `importmap-rails` gems per their instructions

- Move the activeadmin tailwind config to same location `tailwindcss-rails` puts theirs:

        mv tailwind-active_admin.config.js config/tailwind-active_admin.config.js

- Make small alteration to `tailwind-active_admin.config.js` to get the plugin path correct

    ```javascript
    const execSync = require('child_process').execSync;
    const activeAdminPath = execSync('bundle show activeadmin', { encoding: 'utf-8' }).trim();
    
    module.exports = {
      content: [
        `${activeAdminPath}/vendor/javascript/flowbite.js`,
        `${activeAdminPath}/plugin.js`,
        `${activeAdminPath}/app/views/**/*.{arb,erb,html,rb}`,
        './app/admin/**/*.{arb,erb,html,rb}',
        './app/views/active_admin/**/*.{arb,erb,html,rb}',
        './app/views/admin/**/*.{arb,erb,html,rb}',
        './app/views/layouts/active_admin*.{erb,html}',
        './app/javascript/**/*.js'
      ],
      darkMode: "selector",
      plugins: [
        // Need to correct path to plugin.js
        // require(`@activeadmin/activeadmin/plugin`)
        require(`${activeAdminPath}/plugin`)
      ]
    }
   ```
   
- `tailwindcss-rails` has a few different ways to compile the tailwind css. I wanted to use the puma plugin, so need to override the command the puma-plugin uses so it would compile the ActiveAdmin tailwind files, rather than the tailwindcss-rails default.

    Presumably if you wanted both the default tailwindcss and activeadmin's tailwindcss, you'd do this a bit different - maybe the foreman Procfile.dev method would work better. But this worked for the puma plugin:

    Create this initializer:
    
    ```ruby
    module TailwindCommandsOverride
      def compile_command(debug: false, **kwargs)
        command = [
          Tailwindcss::Ruby.executable(**kwargs),
          "-i", Rails.root.join("app/assets/stylesheets/active_admin.css").to_s,
          "-o", Rails.root.join("app/assets/builds/active_admin.css").to_s,
          "-c", Rails.root.join("config/tailwind-active_admin.config.js").to_s
        ]
    
        command << "--minify" unless debug || rails_css_compressor?
    
        postcss_path = Rails.root.join("config/postcss.config.js")
        command += [ "--postcss", postcss_path.to_s ] if File.exist?(postcss_path)
    
        command
      end
    end
    
    module Tailwindcss
      module Commands
        class << self
          prepend TailwindCommandsOverride
        end
      end
    end
    ```
    
    ```ruby
    # config/puma.rb
    plugin :tailwindcss if ENV.fetch("RAILS_ENV", "development") == "development"
    ```

- The ActiveAdmin import maps was a bit convoluted. First pin the package:

    ```
    bin/importmap pin @activeadmin/activeadmin
    ```

- But the version of poppover downloaded doesn't include all the ESM modules, so need to download a different version that does. Also note the `+` character in the filename mucks the importmap up, so that needs to be renamed too:

    ```
    bin/importmap pin @popperjs/core@2.11.8/+esm  --from jsdelivr
    rm vendor/javascript/@popperjs--core.js
    mv vendor/javascript/@popperjs--core--+esm.js vendor/javascript/@popperjs--core.js
    ```

- For some reason poppover doesn't load with the rest of the other importmaps when starting ActiveAdmin, so need to also tell activeadmin to explicitly load it, put this in an initializer

    ```
    ActiveAdmin.importmap.draw do
      pin "@popperjs/core", to: "@popperjs--core.js" # @2.11.8
    end
    ```
    
After that, I have a working ActiveAdmin installation, using `tailwindcss-rails` and `importmap-rails`.  Hopefully the process could be simplfied in the future!

**Reply by amkisko** [reply]
What's the reason to pin `@activeadmin/activeadmin` to main importmap? They are not used in ActiveAdmin layout as far as I understand.

**Reply by Fire-Dragon-DoL** [reply]
I would love some guidance too. I'm trying to use importmaps and my head hurts, for simplifying things it looks actually way more complicated.

**Reply by Fire-Dragon-DoL** [reply]
OK, I think I got it. [ActiveAdmin uses its own defined importmap](https://github.com/activeadmin/activeadmin/blob/v4.0.0.beta15/config/importmap.rb) which the user can't really set things to.
This [importmap is used in the main activeadmin layout](https://github.com/activeadmin/activeadmin/blob/v4.0.0.beta15/app/views/active_admin/_html_head.html.erb#L13), I suspect this is the place where user can add more custom JS.

Adding `pin "@activeadmin/activeadmin", to: "@activeadmin--activeadmin.js" # @4.0.0-beta15` to your own `config/importmap.rb` file **is a mistake that can/will load incorrect nested packages versions**. This is why my `flowbite` was trying to load `@popperjs/core`, because it was an incorrect version, replacing the activeadmin one.

The solution was easy:
- **Do not put anything in your own importmap.rb**
- Do install `@activeadmin/activeadmin` package in `package.json`, you need that **for CSS**, not for JS

**Reply by amkisko** [reply]
A little corrections:

> Do not put anything in your own importmap.rb

You don't have to put anything for ActiveAdmin in `config/importmap.rb`, ActiveAdmin has own importmap definition, you can find hints here in this thread.

> Do install @activeadmin/activeadmin package in package.json, you need that for CSS, not for JS

You don't have to do this, you can fully avoid nodejs usage for ActiveAdmin v4.

Also you can append ActiveAdmin importmap with things you want to have there available, e.g. echarts or other javascript files/libraries.

**Comment by Micka33** [comment]
I have read through several conversations here about that subject, but I could figure it out.

My app was created with `rails new . --database=postgresql --css=tailwind`.  So it uses `propshaft`.

Before Rails 8 (with sprocket), I could do the following to have ActiveAdmin work with Tailwind + SCSS
- In Gemfile
```ruby
...
gem "activeadmin", "~> 3.2"
gem "sass-rails", "~> 6.0"
```
- `active_admin.scss` was moved to `vendor/assets/stylesheets` (can't remember why).
- In `config/initializers/sassc_monkeypatch_for_tailwind.rb`
```ruby
module SassCompressorMonkeypatch
  def call_processor(processor, input)
    super
  rescue SassC::SyntaxError => e
    raise unless processor == Sprockets::SassCompressor
    puts "Warning: Could not compress #{input[:filename]} with Sprockets::SassCompressor. Returning uncompressed result." unless Rails.env.test?
    metadata = (input[:metadata] || {}).dup
    metadata[:data] = input[:data]
    metadata
  end
end
Sprockets::Base.prepend(SassCompressorMonkeypatch)
```

With Rails 8.0.0 (with propshaft), It seems I have to use the beta and I saw [activeadmin_assets](https://github.com/jaynetics/activeadmin_assets).
- In Gemfile
```
gem 'activeadmin', '~> 4.0.0.beta7'
gem 'activeadmin_assets'
```

I might have done it wrong, but that didn't work out. So for now, I have no Idea how to use ActiveAdmin with Rails 8.0.0 (+ propshaft).  :(

Is there a resource written somewhere that could guide me through the steps?

**Reply by jaynetics** [reply]
hi! author of activeadmin_assets here. its not an official solution, so any issues are best posted in its repo. that being said, it works fine in a fresh rails 8 app, with or without propshaft, and with or without `--css=tailwind`. maybe you need to remove special asset setup or asset-related patches.

**Reply by amkisko** [reply]
There are multiple working solutions and options mentioned in this discussion, just select one of and follow it. Especially if you start from scratch, just follow the instructions from the beginning. I use the one I created (which works for Rails 8 too): https://gist.github.com/amkisko/c704c1a6462d573dfa4820ae07d807a6 -- there are things that I would already improve, but it does not matter for now.

**Comment by rafaelsales** [comment]
Rails developer for 10+ years here. I just want to express that it's 2025 and I'm struggling to set this up.

Rails is going in the direction of simplifying things, but active admin v4 seems to be going in the opposite direction. Adding a dependency on `cssbundling-rails` doesn't feel great.

**Reply by henrikbjorn** [reply]
Just use `tailwindcss-rails` and the instructions above with the additional rake tasks. Works great for 5+ projects I have done it on.

Otherwise if you are willing, then `vite_rails` is great.

**Reply by tagliala** [reply]
Side note: `tailwindcss-rails` `v3`, the latest released version will install v4, which is not yet compatible with Active Admin. Ref #8621

**Reply by henrikbjorn** [reply]
@tagliala Add `tailwindcss-ruby` with the correct constraint `~> 3.4.0`

TailwindcssRuby versions follow the released tailwind versions.

**Reply by amkisko** [reply]
Struggling here mostly related to lack of understanding how importmap and tailwindcss bundling works with Rails, once you get it - no more struggling. Basically studying/reading propshaft, importmap, tailwindcss-rails sources will help a lot. ActiveAdmin v4 does not do anything in addition.

I confirm too that "node-less" setup works, I went through it in multiple projects (some with fresh v4, some migrated from v3) which are already in production with real users.

**Comment by Olzhassss** [comment]
In case someone tries to set it up with Tailwindcss v4, `tailwind-active_admin.config.js` can be migrated to `app/assets/stylesheets/active_admin.css` as follows (given bundle gems location at `vendor/bundle/ruby/xx/gems`):

```
@import 'tailwindcss' source(none);
@source '../../../vendor/bundle/ruby/3.4.0/gems/activeadmin-4.0.0.beta15/vendor/javascript/flowbite.js';
@source '../../../vendor/bundle/ruby/3.4.0/gems/activeadmin-4.0.0.beta15/app/views';
@plugin '../../../vendor/bundle/ruby/3.4.0/gems/activeadmin-4.0.0.beta15/plugin.js';
@source '../../../vendor/bundle/ruby/3.4.0/gems/activeadmin-4.0.0.beta15/plugin.js';
/* add other @source tags from tailwind-active_admin.config.js if necessary */
```

As you see, I used ruby 3.4.0 and activeadmin-4.0.0.beta15. After this, the js config file can be deleted.
`active_admin.rake` from https://medium.com/@henrikbjorn/til-how-to-use-activeadmin-v4-without-installing-nodejs-ccdaf3ace005 can also be slightly updated by removing `-c` arg for both tasks.

It is not useful to add `@source` to folders in `app/` because, when using tailwindcss v4 with rails, tailwindcss scans most files in the project already, including all of `app/`.

**Reply by Olzhassss** [reply]
Also I noticed that sourcing plugin.js fails because tailwindcss v4 encounters unknown class `ring-opacity-5`, so this class would have to be manually removed from the file for this to work.

**Reply by amkisko** [reply]
ActiveAdmin tailwind v3 plugin is not compatible with tailwind v4, it requires full rewrite.

Partially working example available here: https://gist.github.com/amkisko/af1b2f7dc4f0f941437ea16400277864 (you may use any genai-powered editor to complete this and even maybe contribute if you want)
